File: src/main/java/com/github/caijh/commons/util/Strings.java
package com.github.caijh.commons.util;

import java.math.BigDecimal;
import java.text.NumberFormat;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Objects;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import com.github.caijh.commons.util.constants.Delimiters;
import org.apache.commons.lang3.StringUtils;


public class Strings extends org.apache.commons.lang3.StringUtils {

    private Strings() {

    }

    public static int[] toIntArray(String s, String regex) {
        if (StringUtils.isBlank(s)) {
            return new int[]{};
        }

        return Arrays.stream(s.split(regex)).filter(Strings::isNotBlank).mapToInt(Integer::valueOf).toArray();
    }

    /**
     * @param s     字符串
     * @param regex 分割符
     * @return list of string
     */
    public static List<String> toList(String s, String regex) {
        if (Objects.isNull(s)) {
            return Collections.emptyList();
        }
        return Arrays.stream(s.split(regex)).filter(Strings::isNotBlank).collect(Collectors.toList());
    }

    /**
     * @param s      字符串
     * @param regex  分割符
     * @param mapper 转化函数
     * @param <R>    list中元素的类型
     * @return list of R
     */
    public static <R> List<R> toList(String s, String regex, Function<String, R> mapper) {
        if (Objects.isNull(s)) {
            return Collections.emptyList();
        }
        return Arrays.stream(s.split(regex)).filter(Strings::isNotBlank).map(mapper).collect(Collectors.toList());
    }

    public static List<String> toListByComma(String s) {
        return toList(s, Delimiters.COMMA);
    }

    public static List<Integer> toIntListByComma(String s) {
        return toList(s, Delimiters.COMMA, Strings::toInt);
    }

    public static long toLong(String s) {
        return Long.parseLong(s.trim());
    }

    public static int toInt(String s) {
        return Integer.parseInt(s.trim());
    }

    public static String firstCharUpperCase(String str) {
        if (isBlank(str)) {
            throw new IllegalArgumentException();
        }
        StringBuilder stringBuilder = new StringBuilder(str);
        stringBuilder.setCharAt(0, Character.toUpperCase(str.charAt(0)));
        return stringBuilder.toString();
    }

    public static boolean isAnyBlank(String... s) {
        return Stream.of(s).anyMatch(Strings::isBlank);
    }

    /**
     * Supports rendering of Java numeric types float, double,
     * and BigDecimal in "default" format and in format that
     * avoids use of scientific notation.
     */
    private enum Format {
        DEFAULT {
            @Override
            public String fromFloat(final float floatValue) {
                return numberFormat.format(floatValue);
            }

            @Override
            public String fromDouble(final double doubleValue) {
                return numberFormat.format(doubleValue);
            }

            @Override
            public String fromBigDecimal(final BigDecimal bigDecimalValue) {
                return bigDecimalValue.toPlainString();
            }
        };

        private static final NumberFormat numberFormat = NumberFormat.getInstance();

        static {
            numberFormat.setMaximumFractionDigits(Integer.MAX_VALUE);
            numberFormat.setGroupingUsed(false);
        }

        public abstract String fromFloat(final float floatValue);

        public abstract String fromDouble(final double doubleValue);

        public abstract String fromBigDecimal(final BigDecimal bigDecimalValue);
    }

    public static String toString(Object object) {
        if (object instanceof Date) {
            return DateUtils.format((Date) object);
        }
        if (object instanceof Number) {
            if (object instanceof Float) {
                return Format.DEFAULT.fromFloat((Float) object);
            }
            if (object instanceof Double) {
                return Format.DEFAULT.fromDouble((Double) object);
            }
            if (object instanceof BigDecimal) {
                return Format.DEFAULT.fromBigDecimal((BigDecimal) object);
            }
        }
        return object.toString();
    }

}

File: src/main/kotlin/com/github/caijh/commons/util/Strings.kt
package com.github.caijh.commons.util

import com.github.caijh.commons.util.Collections.emptyList
import org.apache.commons.lang3.StringUtils
import java.util.*
import java.util.function.Function
import java.util.regex.Pattern
import java.util.stream.Collectors
import java.util.stream.Stream

object Strings {

    const val EMPTY = ""

    /**
     * 将字符串转换为int array.
     * 如1,2,3,4,5, regex：,
     */
    @JvmStatic
    fun toIntArray(s: String, regex: String): IntArray {
        return if (StringUtils.isBlank(s)) {
            intArrayOf()
        } else Arrays.stream(s.split(regex.toRegex()).toTypedArray())
            .filter { cs: String -> StringUtils.isNotBlank(cs) && StringUtils.isNumeric(cs) }
            .mapToInt { num: String -> Integer.valueOf(num) }
            .toArray()
    }

    /**
     * @param s     字符串
     * @param regex 分割符
     * @return list of string
     */
    @JvmStatic
    fun toList(s: String, regex: String): List<String> {
        return if (Objects.isNull(s)) {
            emptyList()
        } else Arrays.stream(s.split(regex.toRegex()).toTypedArray())
            .filter { cs: String? -> StringUtils.isNotBlank(cs) }
            .collect(Collectors.toList())
    }

    /**
     * @param s      字符串
     * @param regex  分割符
     * @param mapper 转化函数
     * @param <R>    list中元素的类型
     * @return list of R
    </R> */
    @JvmStatic
    fun <R> toList(s: String, regex: String, mapper: Function<String, R>): List<R> {
        return if (Objects.isNull(s)) {
            emptyList()
        } else Arrays.stream(s.split(regex.toRegex()).toTypedArray())
            .filter { cs: String? -> StringUtils.isNotBlank(cs) }
            .map(mapper).collect(Collectors.toList())
    }

    @JvmStatic
    fun toListByComma(s: String): List<String> {
        return toList(s, Delimiters.COMMA)
    }

    @JvmStatic
    fun toIntListByComma(s: String): List<Int> {
        return toList(s, Delimiters.COMMA) { obj: String -> toInt(obj) }
    }

    @JvmStatic
    fun toLong(s: String): Long {
        return s.trim { it <= ' ' }.toLong()
    }

    @JvmStatic
    fun isInt(s: String): Boolean {
        val pattern = Pattern.compile("^[1-9][0-9]*$")
        val matcher = pattern.matcher(s)
        return matcher.matches()
    }

    @JvmStatic
    fun toInt(s: String): Int {
        return s.trim { it <= ' ' }.toInt()
    }

    @JvmStatic
    fun firstCharUpperCase(str: String): String {
        require(!StringUtils.isBlank(str))
        val stringBuilder = StringBuilder(str)
        stringBuilder.setCharAt(0, Character.toUpperCase(str[0]))
        return stringBuilder.toString()
    }

    @JvmStatic
    fun firstCharLowerCase(str: String): String {
        require(!StringUtils.isBlank(str))
        val stringBuilder = StringBuilder(str)
        stringBuilder.setCharAt(0, Character.toLowerCase(str[0]))
        return stringBuilder.toString()
    }

    @JvmStatic
    fun isAnyBlank(vararg s: String?): Boolean {
        return Stream.of(*s).anyMatch { cs: String? -> StringUtils.isBlank(cs) }
    }

    /**
     * 判断字符串是否为空
     *
     * @param s 字符串
     * @return true, if string is blank or null
     */
    @JvmStatic
    fun isBlank(s: String?): Boolean {
        return StringUtils.isBlank(s)
    }

    /**
     * 判断字符串是否不为空
     */
    @JvmStatic
    fun isNotBlank(s: String?): Boolean {
        return StringUtils.isNotBlank(s)
    }

}
